### PR TITLE
Introduce changesets for CHANGELOG auto-generation

### PR DESCRIPTION
**Pull request by JounQin**
> Which issue, if any, is this issue related to?

Related #6253

> Is there anything in the PR that needs further explanation?

No, it's self-explanatory.

### CODE REVIEW

File: .github/workflows/release.yml
Comment (JounQin) [inline .github/workflows/release.yml:31]: This workflow only affects CHANGELOG for now.

File: package.json
Comment (ybiquitous) [inline package.json:112]: [suggest] Should the new packages be in `devDependencies`?

Comment (JounQin) [reply package.json:112]: Sorry, I forgot to use `-d` flag.

File: .changeset/config.json
Comment (ybiquitous) [inline .changeset/config.json:4]: [ask] Does `@changesets/changelog-github` match our existing CHANGELOG format? Isn't a custom formatter necessary?

Comment (JounQin) [reply .changeset/config.json:4]: The default format is different.

From https://github.com/stylelint/stylelint/issues/6253#issuecomment-1216039310

> We can update CHANGELOG.md automatically via the two things above, right? Additionally, do we need to write a custom formatter to keep the existing CHANGELOG.md format?

If we want to keep the existing CHANGELOG.md format, then we need to write a custom formatter.

Comment (ybiquitous) [reply .changeset/config.json:4]: I see. I think keeping the format is better because it will be easier if we revert the changes.

Is writing a custom formatter hard (needs many lines of code)?

Comment (JounQin) [reply .changeset/config.json:4]: > Is writing a custom formatter hard (needs many lines of code)?

In my imagine, it should not be tough, I'll add a custom one in this PR later.

File: .github/workflows/release.yml
Comment (ybiquitous) [inline .github/workflows/release.yml:1]: [suggest] Considering consistency with other workflows, `...ing` seems better:

```suggestion
name: Releasing
```

And `release.yml` → `releasing.yml`.

File: .github/workflows/release.yml
Comment (ybiquitous) [inline .github/workflows/release.yml:7]: [ask] According to the [`changesets/action` README](https://github.com/changesets/action#without-publishing), `concurrency` is set in the example. Do we also need it?

```suggestion

concurrency: ${{ github.workflow }}-${{ github.ref }}

```

File: .github/workflows/release.yml
Comment (ybiquitous) [inline .github/workflows/release.yml:22]: [suggest] Does it make sense to use LTS for less maintenance?

```suggestion
      - name: Set up Node.js LTS
        uses: actions/setup-node@v3
        with:
          node-version: 'lts/*'
```

File: .github/workflows/release.yml
Comment (ybiquitous) [inline .github/workflows/release.yml:26]: [suggest] What about making deps installation consistent with other workflows?

```suggestion
      - name: Install latest npm 
        run: npm install --global npm@latest 
  
      - name: Install dependencies 
        run: npm ci 
```

https://github.com/stylelint/stylelint/blob/2c2a57cdca138aa6afe225b90b3fa6d2110245fc/.github/workflows/testing.yml#L41-L45

File: .github/workflows/release.yml
Comment (ybiquitous) [inline .github/workflows/release.yml:33]: [suggest] Let's simplify the change and keep the commit/PR title consistent with old commits. We can change them when we will automate publishing.
It's ideal for including a version number like `Prepare 14.10.01`, but `Prepare release` seems not bad to me.

```suggestion
      - name: Create release pull request
        uses: changesets/action@v1
        with:
          commit: 'Prepare release'
          title: 'Prepare release'
```

**Review by ybiquitous — CHANGES_REQUESTED** [review]

File: .github/workflows/release.yml
Comment (JounQin) [reply .github/workflows/release.yml:7]: Perhaps we should align with the official document, I've never used this previously.

File: .github/workflows/release.yml
Comment (JounQin) [reply .github/workflows/release.yml:22]: 👍

File: .github/workflows/release.yml
Comment (JounQin) [reply .github/workflows/release.yml:26]: 👍

File: .github/workflows/release.yml
Comment (JounQin) [reply .github/workflows/release.yml:33]: I want `Prepare 14.10.01` (with version) too, but haven't find a good solution.

File: .github/workflows/release.yml
Comment (JounQin) [reply .github/workflows/release.yml:1]: Sure.

File: .changeset/changelog-stylelint.js
Comment (JounQin) [inline .changeset/changelog-stylelint.js:11]: Sections like `### Patch Changes` can not be omitted for now.

File: .changeset/changelog-stylelint.js
Comment (JounQin) [inline .changeset/changelog-stylelint.js:86]: Do we want `thanks` mentions included in this PR?

**Comment by JounQin** [comment]
I'm going to migrate current HEAD changelog into changesets in this PR, HDYT? @ybiquitous

**Review by ybiquitous — COMMENTED** [review]
@JounQin Thank you. I've left some comments, but it seems there are no problems.

File: .changeset/changelog-stylelint.js
Comment (ybiquitous) [reply .changeset/changelog-stylelint.js:11]: I think we can tolerate it.

**Comment by ybiquitous** [comment]
[Remind] If this new workflow fits, we'll need to update the maintainer guide after this PR is merged:

- https://github.com/stylelint/stylelint/blob/613334b7711654bba5a4f75f01780131757f0656/docs/maintainer-guide/pull-requests.md?plain=1#L27
- https://github.com/stylelint/stylelint/blob/613334b7711654bba5a4f75f01780131757f0656/docs/maintainer-guide/releases.md?plain=1#L8-L9

File: .changeset/changelog-stylelint.js
Comment (ybiquitous) [reply .changeset/changelog-stylelint.js:86]: We can reconsider this point later (when updating the changelog).

**Review by ybiquitous — APPROVED** [review]
I'd like to move forward with the evaluation of new workflow, so LGTM 👍🏼

**Comment by JounQin** [comment]
I'll wait for one more day to see if there will be any new reviews, then I'm going to merge. 🍺

**Comment by ybiquitous** [comment]
> Just found I haven't finished this part: [#6274 (comment)](https://github.com/stylelint/stylelint/pull/6274#issuecomment-1217915299)

Sorry, I missed it. Can you push the change into this branch, please?

**Comment by JounQin** [comment]
> > Just found I haven't finished this part: [#6274 (comment)](https://github.com/stylelint/stylelint/pull/6274#issuecomment-1217915299)
> 
> Sorry, I missed it. Can you push the change into this branch, please?

Sure, I'd like to finish them today.

**Review by jeddy3 — COMMENTED** [review]
Looking good. Let's adopt the [ESLint format](https://github.com/eslint/eslint/blob/main/CHANGELOG.md) for mentions because it's simple and more in line with the existing format of the changelog.

For example:

- Fixed: `time-min-milliseconds` end positions ([#6273](https://github.com/stylelint/stylelint/pull/6273)) (@ybiquitous).

File: .changeset/calm-tables-hide.md
Comment (jeddy3) [inline .changeset/calm-tables-hide.md:5]:
```suggestion
Fixed: `selector-pseudo-class-allowed-list` end positions and message
```

Should these messages not include the pull request number as that (and the mention) is automatically added by the script?

Comment (JounQin) [reply .changeset/calm-tables-hide.md:5]: In this PR, it can't, because the changesets are generated in this PR. 🤣

Comment (jeddy3) [reply .changeset/calm-tables-hide.md:5]: > In this PR, it can't, because the changesets are generated in this PR. 🤣

Oh, I see. Makes sense.

> Do you want me to add the mentions manually in this PR?

Narh, it's not worth it if it's manual.  They can start appearing next time around when it's automated.

**Comment by ybiquitous** [comment]
Ah, sorry. I've updated the changelog for #6275, and conflicts occur. Could you please resolve the conflicts? @JounQin

**Comment by JounQin** [comment]
> Ah, sorry. I've updated the changelog for #6275, and conflicts occur. Could you please resolve the conflicts? @JounQin

Sure, wait a minute.

**Review by ybiquitous — APPROVED** [review]
Thanks! LGTM 👍🏼

**Comment by JounQin** [comment]
@jeddy3 Are we good to merge right now?

**Review by jeddy3 — APPROVED** [review]
LGTM, thank you!

**Comment by JounQin** [comment]
https://github.com/stylelint/stylelint/runs/7897522989?check_suite_focus=true#step:6:61

@ybiquitous @jeddy3 It seems the permissions are not working as expected, would you like to help?

**Pull request merged**
